The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10763575981/job/29845251292) [(plain)](https://github.com/rust-lang/rust/commit/64fa312fff7b2bbe9f11ec4551cbda0d56879fd1/checks/29845251292/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.734 Building wheels for collected packages: reuse
#13 2.735   Building wheel for reuse (pyproject.toml): started
#13 2.981   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 2.982   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 2.982   Stored in directory: /tmp/pip-ephem-wheel-cache-n_7p604t/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 2.985 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.390 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.391 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.928 Collecting virtualenv
#13 3.928 Collecting virtualenv
#13 3.966   Downloading virtualenv-20.26.4-py3-none-any.whl (6.0 MB)
#13 4.047      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 76.1 MB/s eta 0:00:00
#13 4.107 Collecting platformdirs<5,>=3.9.1
#13 4.111   Downloading platformdirs-4.3.2-py3-none-any.whl (18 kB)
#13 4.152 Collecting filelock<4,>=3.12.2
#13 4.179 Collecting distlib<1,>=0.3.7
#13 4.183   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.189      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 98.5 MB/s eta 0:00:00
#13 4.189      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 98.5 MB/s eta 0:00:00
#13 4.270 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.460 Successfully installed distlib-0.3.8 filelock-3.16.0 platformdirs-4.3.2 virtualenv-20.26.4
#13 DONE 4.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      221120 kB
DirectMap2M:     9216000 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/adf8d168af9334a8bf940824fcf4207d01e05ae5/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-adf8d168af9334a8bf940824fcf4207d01e05ae5-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
##[endgroup]
fmt check
Diff in /checkout/src/tools/build_helper/src/git.rs:167:
 ///
 /// This can result in formatting thousands of files instead of a dozen,
 /// so we should warn the user something is wrong.
-pub fn warn_old_master_branch(
-    config: &GitConfig<'_>,
-    git_dir: &Path,
-) {
+pub fn warn_old_master_branch(config: &GitConfig<'_>, git_dir: &Path) {
     if crate::ci::CiEnv::is_ci() {
         // this warning is useless in CI,
         // and CI probably won't have the right branches anyway.
Diff in /checkout/src/tools/build_helper/src/git.rs:180:
     let mut updated_master = "the upstream master branch".to_string();
     match warn_old_master_branch_(config, git_dir, &mut updated_master) {
         Ok(branch_is_old) => {
-            if !branch_is_old { return }
+            if !branch_is_old {
+            }
             // otherwise fall through and print the rest of the warning
         }
         Err(err) => {
         Err(err) => {
Diff in /checkout/src/tools/build_helper/src/git.rs:210:
                 return Err(Box::new(err));
             }
             std::fs::metadata(Path::new(&gcd).join("refs/remotes").join(&updated_master))?
+        }
     };
     };
     if meta.modified()?.elapsed()? > WARN_AFTER {
         eprintln!("warning: {updated_master} has not been updated in 10 days");
Diff in /checkout/src/tools/build_helper/src/git.rs:221:
 
 
 fn git_common_dir(dir: &Path) -> Result<String, String> {
-    output_result(Command::new("git")
-                  .arg("-C").arg(dir)
-                  .arg("rev-parse").arg("--git-common-dir"))
+    output_result(Command::new("git").arg("-C").arg(dir).arg("rev-parse").arg("--git-common-dir"))
         .map(|x| x.trim().to_string())
 
 
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_query_system/src/query/plumbing.rs" "/checkout/compiler/rustc_query_system/src/query/config.rs" "/checkout/compiler/rustc_query_system/src/query/mod.rs" "/checkout/compiler/rustc_query_system/src/ich/hcx.rs" "/checkout/compiler/rustc_query_system/src/ich/impls_syntax.rs" "/checkout/compiler/rustc_query_system/src/ich/mod.rs" "/checkout/compiler/rustc_query_system/src/lib.rs" "/checkout/compiler/rustc_query_system/src/values.rs" "/checkout/compiler/rustc_query_system/src/dep_graph/dep_node.rs" "/checkout/compiler/rustc_query_system/src/dep_graph/serialized.rs" "/checkout/compiler/rustc_query_system/src/dep_graph/query.rs" "/checkout/compiler/rustc_query_system/src/dep_graph/debug.rs" "/checkout/compiler/rustc_query_system/src/dep_graph/mod.rs" "/checkout/compiler/rustc_query_system/src/dep_graph/graph.rs" "/checkout/compiler/rustc_query_system/src/dep_graph/edges.rs" "/checkout/compiler/rustc_error_codes/src/lib.rs" "/checkout/src/tools/compiletest/src/json.rs" "/checkout/src/tools/compiletest/src/tests.rs" "/checkout/src/tools/compiletest/src/read2/tests.rs" "/checkout/src/tools/compiletest/src/compute_diff.rs" "/checkout/src/tools/compiletest/src/raise_fd_limit.rs" "/checkout/src/tools/compiletest/src/runtest.rs" "/checkout/src/tools/compiletest/src/errors.rs" "/checkout/src/tools/compiletest/src/lib.rs" "/checkout/src/tools/compiletest/src/util.rs" "/checkout/src/tools/compiletest/src/read2.rs" "/checkout/src/tools/compiletest/src/errors/tests.rs" "/checkout/src/tools/compiletest/src/common.rs" "/checkout/src/tools/compiletest/src/runtest/tests.rs" "/checkout/src/tools/compiletest/src/runtest/coverage.rs" "/checkout/src/tools/compiletest/src/runtest/debugger.rs" "/checkout/src/tools/compiletest/src/header.rs" "/checkout/src/tools/compiletest/src/util/tests.rs" "/checkout/src/tools/compiletest/src/command-list.rs" "/checkout/src/tools/compiletest/src/header/tests.rs" "/checkout/src/tools/compiletest/src/header/cfg.rs" "/checkout/src/tools/compiletest/src/header/needs.rs" "/checkout/src/tools/compiletest/src/header/test-auxillary/known_directive.rs" "/checkout/src/tools/compiletest/src/header/test-auxillary/unknown_directive.rs" "/checkout/src/tools/compiletest/src/header/test-auxillary/known_legacy_directive.rs" "/checkout/src/tools/compiletest/src/header/test-auxillary/error_annotation.rs" "/checkout/src/tools/compiletest/src/main.rs" "/checkout/src/tools/unstable-book-gen/src/main.rs" "/checkout/src/tools/wasm-component-ld/src/main.rs" "/checkout/src/tools/error_index_generator/main.rs" "/checkout/src/tools/build_helper/src/stage0_parser.rs" "/checkout/src/tools/build_helper/src/ci.rs" "/checkout/src/tools/build_helper/src/git.rs" "/checkout/src/tools/build_helper/src/lib.rs" "/checkout/src/tools/build_helper/src/metrics.rs" "/checkout/src/tools/build_helper/src/util.rs" "/checkout/src/tools/build_helper/src/drop_bomb/tests.rs" "/checkout/src/tools/build_helper/src/drop_bomb/mod.rs" "/checkout/src/tools/remote-test-client/src/main.rs" "/checkout/src/tools/miropt-test-tools/src/lib.rs" "/checkout/src/tools/linkchecker/tests/checks.rs" "/checkout/src/tools/linkchecker/main.rs" "/checkout/src/tools/rustbook/src/main.rs" "/checkout/src/tools/rustdoc-gui-test/src/config.rs" "/checkout/src/tools/rustdoc-gui-test/src/main.rs" "/checkout/src/tools/rust-installer/src/combiner.rs" "/checkout/src/tools/rust-installer/src/lib.rs" "/checkout/src/tools/rust-installer/src/scripter.rs" "/checkout/compiler/rustc_query_system/src/query/caches.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Sun Sep  8 22:30:54 UTC 2024
  network time: Sun, 08 Sep 2024 22:30:54 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/src/bootstrap/src/core/sanity.rs
+++ b/src/bootstrap/src/core/sanity.rs
@@ -378,13 +378,5 @@ $ pacman -R cmake && pacman -S mingw-w64-x86_64-cmake
         cmd_finder.must_have(s);
     }
 
-    // this warning is useless in CI,
-    // and CI probably won't have the right branches anyway.
-    if !build_helper::ci::CiEnv::is_ci() {
-        if let Err(e) = warn_old_master_branch(&build.config.git_config(), &build.config.src)
-            .map_err(|e| e.to_string())
-        {
-            eprintln!("unable to check if upstream branch is old: {e}");
-        }
-    }
+    warn_old_master_branch(&build.config.git_config(), &build.config.src);
 }

--- a/src/tools/build_helper/src/git.rs
+++ b/src/tools/build_helper/src/git.rs
@@ -167,21 +167,20 @@ pub fn get_git_untracked_files(
 ///
 /// This can result in formatting thousands of files instead of a dozen,
 /// so we should warn the user something is wrong.
-pub fn warn_old_master_branch(
-    config: &GitConfig<'_>,
-    git_dir: &Path,
-) -> Result<(), Box<dyn std::error::Error>> {
-    use std::time::Duration;
-    const WARN_AFTER: Duration = Duration::from_secs(60 * 60 * 24 * 10);
-    let updated_master = updated_master_branch(config, Some(git_dir))?;
-    let branch_path = git_dir.join(".git/refs/remotes").join(&updated_master);
-    match std::fs::metadata(branch_path) {
-        Ok(meta) => {
-            if meta.modified()?.elapsed()? > WARN_AFTER {
-                eprintln!("warning: {updated_master} has not been updated in 10 days");
-            } else {
-                return Ok(());
+pub fn warn_old_master_branch(config: &GitConfig<'_>, git_dir: &Path) {
+    if crate::ci::CiEnv::is_ci() {
+        // this warning is useless in CI,
+        // and CI probably won't have the right branches anyway.
+        return;
+    }
+    // this will be overwritten by the actual name, if possible
+    let mut updated_master = "the upstream master branch".to_string();
+    match warn_old_master_branch_(config, git_dir, &mut updated_master) {
+        Ok(branch_is_old) => {
+            if !branch_is_old {
+                return;
             }
+            // otherwise fall through and print the rest of the warning
         }
         Err(err) => {
             eprintln!("warning: unable to check if {updated_master} is old due to error: {err}")
@@ -189,7 +188,38 @@ pub fn warn_old_master_branch(
     }
     eprintln!(
         "warning: {updated_master} is used to determine if files have been modified\n\
-               warning: if it is not updated, this may cause files to be needlessly reformatted"
+         warning: if it is not updated, this may cause files to be needlessly reformatted"
     );
-    Ok(())
+}
+
+pub fn warn_old_master_branch_(
+    config: &GitConfig<'_>,
+    git_dir: &Path,
+    updated_master: &mut String,
+) -> Result<bool, Box<dyn std::error::Error>> {
+    use std::time::Duration;
+    *updated_master = updated_master_branch(config, Some(git_dir))?;
+    let branch_path = git_dir.join(".git/refs/remotes").join(&updated_master);
+    const WARN_AFTER: Duration = Duration::from_secs(60 * 60 * 24 * 10);
+    let meta = match std::fs::metadata(&branch_path) {
+        Ok(meta) => meta,
+        Err(err) => {
+            let gcd = git_common_dir(&git_dir)?;
+            if branch_path.starts_with(&gcd) {
+                return Err(Box::new(err));
+            }
+            std::fs::metadata(Path::new(&gcd).join("refs/remotes").join(&updated_master))?
+        }
+    };
+    if meta.modified()?.elapsed()? > WARN_AFTER {
+        eprintln!("warning: {updated_master} has not been updated in 10 days");
+        Ok(true)
+    } else {
+        Ok(false)
+    }
+}
+
+fn git_common_dir(dir: &Path) -> Result<String, String> {
+    output_result(Command::new("git").arg("-C").arg(dir).arg("rev-parse").arg("--git-common-dir"))
+        .map(|x| x.trim().to_string())
 }

